text/3503-frontmatter.md

```diff
@@ -0,0 +1,694 @@
+We may at least defer stabilization of infostrings.
+
+The infostring syntax was selected to allow file names (e.g. `Cargo.lock`).
+Additional attributes are left to a future possibility.
+
+## Syntax
+
+[RFC 3502] lays out some design principles, including
+- Single-file packages should have a first-class experience
+  - Provides a higher quality of experience (doesn't feel like a hack or tacked on)
+  - Transferable knowledge, whether experience, stackoverflow answers, etc
+  - Easier unassisted migration between single-file and multi-file packages
+  - The more the workflows deviate, the higher the maintenance and support costs for the cargo team
+
+This led us to wanting to re-use the existing manifest format inside of Rust code.
+The question is what that syntax for embedding should be.
+
+When choosing the syntax, our care-abouts are
+- How obvious it is for new users when they see it
+- How easy it is for newer users to remember it and type it out
+- How machine editable it is for `cargo add` and friends
+- Needs to be valid Rust code for quality of error messages, etc
+- Simple enough syntax for tools to parse without a full Rust parser
+  - Leave Rust syntax errors to rustc, rather than masking them with lower quality ones
+  - Ideally we allows random IDE tools (e.g. [`crates.nvim`](https://github.com/Saecki/crates.nvim) to still have easy access to the manifest
+- Leave the door open in case we want to reuse the syntax for embedded lockfiles
+- Leave the door open for single-file `lib`s
+
+### Why add to Rust syntax, rather than letting Cargo handle it
+
+The most naive way for cargo to handle this is for cargo to strip the manifest, write the Rust file to a temporary file, and compile that.
+This is what has traditionally been done with various iterations of "cargo script".
+
+This provides a second-class experience which runs counter to one of the design goals
+- Error messages, `cargo metadata`, etc point to the stripped source with an "odd" path, rather than the real source
+- Every tool that plans to support it would need to be updated to do stripping (`cargo fmt`, `cargo clippy`, etc)
+
+A key part in there is "plan to support".
+We'd need to build up buy-in for tools to be supporting a Cargo-only syntax.
+This becomes more difficult when the tool in question tries to be Cargo-agnostic.
+By having Cargo-agnostic external tool syntax in Rust, this mostly becomes transparent.
+
+We could build a special relationship with rustc to support this.
+For example, rustdoc passes code to rustc on stdin and sets `UNSTABLE_RUSTDOC_TEST_PATH` and `UNSTABLE_RUSTDOC_TEST_LINE` to control how errors are rendered.
+We could then also special case the messages inside of cargo.
+This both adds a major support burden to keep this house of lies standing but still falls short when it comes to tooling support.
+Now every tool that wants to support the Cargo-only syntax has to build their own house of lies.
```

Comment:
Yes, treating it as a container format is one potential direction to go.  Each of the directions comes with different technical trade offs but also potential ecosystem effects that we are trying to predict.

For myself, a cargo-specific container format would create a "second class", "non-native" experience for this.  We'd be going from "this is rust" and any tools that know how to work with Rust Just Work to "this is a special cargo format" and now every tool must be adapted to work with this, every syntax highlighter updated (with a new standardized name for it), etc.  I suspect this will lead to a lot more of patch-work of support within editors, editor extensions, command-line tools, web tools, etc.

For myself, I see having first-class support being able to have a major shift in how people program Rust and communicate about Rust.  This will be viewed as Rust and tacking it on as "this other" format dampens the potential for this.

There is also precedence.  Just recently, Python accepted [PEP 723](https://peps.python.org/pep-0723/) though they went with route of repurposing unstructured comments to have structure as defined by a regex.